### PR TITLE
Ported hair gradients from TGMC.

### PR DESCRIPTION
**Pull request by KathyRyals**
## About The Pull Request

Ported hair gradients from TGMC (credits to @XSlayer300 who ported it from VORE/Polaris). You can currently access them by using the hair spray item (library vendors or the new "Colorist" quirk).

image

Cool colour stuff.

image

Can be used for nice, soft shading.

## Why It's Good For The Game

It looks beautiful. We're all good and we deserve beautiful stuff.

## TODO

- [x] Code the hair spray.
- [x] Get a cool sprite for it.
- [x] Code in the quirk.

## Changelog
:cl: Cimika, with thanks to NamelessFairy for additional patterns and AdipemDragon for the spray icon.
add: Added the ability to dye your hair with gradients by using a hair dye spray.
add: The new Colorist quirk, allowing you to spawn with a hair dye spray.
/:cl:

### CODE REVIEW

**Comment by Serxule** [comment]
does it work with wigs?

**Comment by KathyRyals** [comment]
Nope. Only hair.

**Comment by GoldenAlpharex** [comment]
Just so we're on the same page, this won't _force_ people to use gradients, it's just an additional option, right?

**Comment by KathyRyals** [comment]
Of course ! You can set it to "None".

**Comment by GoldenAlpharex** [comment]
> Of course ! You can set it to "None".

And it's "None" by default?

**Comment by Fikou** [comment]
arent char prefs blocked until the character pref system is fully reworked by mothblocks?

**Comment by Twaticus** [comment]
> arent char pref changes blocked until the character pref system is fully reworked by mothblocks?

yes.

**Comment by Twaticus** [comment]
being reworked from pref to a quirk

**Comment by TheBonded** [comment]
> does it work with wigs?

The real question is, does it work with wi**n**gs?

**Comment by wesoda25** [comment]
Whole new era of snowflakes

**Comment by optimumtact** [comment]
> being reworked from pref to a quirk

I do not think this is appropriate, and should not be done as a workaround to a legitimate pause in operations while prefs are revamped.

I'd like to see the decision behind this

**Comment by KathyRyals** [comment]
Well, I think gradient hair looks very good, a lot of people seem to be impatient to try it and use it on our servers, and I just want to find a way that please everyone and allows for it to be merged. Is there a way you would prefer it to be done ?
The quirk idea I had was a free quirk that makes you spawn with a dye kit that you can use in hand to select a gradient and a colour for your hair.

**Comment by KathyRyals** [comment]
I could also make a machine, like a dying stand or something ! I just want to do something neat for people, and I heard that the pref rework might take a long time to be finished (I'm sorry if that's wrong, but that's my current understanding).

**Comment by uwubli** [comment]
good. give me. i will eat this pr

File: code/__HELPERS/global_lists.dm [hunk 1]
```diff
@@ -46,6 +46,11 @@
 		GLOB.surgeries_list += new path()
 	sortList(GLOB.surgeries_list, /proc/cmp_typepaths_asc)
 
+	// Hair Gradients - Initialise all /datum/sprite_accessory/hair_gradient into an list indexed by gradient-style name
+	for(var/path in subtypesof(/datum/sprite_accessory/hair_gradient))
+		var/datum/sprite_accessory/hair_gradient/H = new path()
```
Comment (Iamgoofball) [inline code/__HELPERS/global_lists.dm:51]: this is a minor nitpick but initial() abuse might be usable here to cut down on news and object refs floating about

File: code/game/objects/items/dyekit.dm [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	if(!new_grad_style)
+		return
+
+	var/new_grad_color = input(usr, "Choose a secondary hair color:", "Character Preference","#"+human_target.grad_color) as color|null
+	if(!new_grad_color)
+		return
+
+	human_target.grad_style = new_grad_style
+	human_target.grad_color = sanitize_hexcolor(new_grad_color)
+	to_chat(human_target, "<span class='notice'>You apply the hair dye...</span>")
```
Comment (Iamgoofball) [inline code/game/objects/items/dyekit.dm:29]: grammar nitpick, should reword this to imply you're starting to take action, not that you took the action, as is the norm for do_after actions

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 3]
```diff
@@ -18,6 +18,10 @@
 	var/hair_color = "000"
 	var/hairstyle = "Bald"
 
+	//Gradient color and style
+	var/grad_color = "000"
+	var/grad_style = "None"
```
Comment (Iamgoofball) [inline code/modules/mob/living/carbon/human/human_defines.dm:23]: these shouldn't be strings but lmao hair code so eh, not much that can be done w/o rewriting most of hair code

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -612,14 +617,28 @@ GLOBAL_LIST_EMPTY(roundstart_races)
...
 						hair_overlay.color = "#" + H.hair_color
+
+					//Gradients
+					grad_style = H.grad_style
+					grad_color = H.grad_color
+					if(grad_style && grad_style != "None")
+						var/datum/sprite_accessory/gradient = GLOB.hair_gradients_list[grad_style]
+						var/icon/temp = new/icon("icon" = gradient.icon, "icon_state" = gradient.icon_state)
+						var/icon/temp_hair = new/icon("icon" = hair_file, "icon_state" = hair_state)
+						temp.Blend(temp_hair, ICON_ADD)
```
Comment (Iamgoofball) [inline code/modules/mob/living/carbon/human/species.dm:628]: cache the result of this for the gradient style, colors, and hair icon used, so as to avoid repeated Blend() calls, which are expensive

**Review by Iamgoofball — CHANGES_REQUESTED** [review]
side note: do dye spray cans need to be separate from normal spray cans?

File: code/game/objects/items/dyekit.dm (hunk 2 above)
Comment (KathyRyals) [reply code/game/objects/items/dyekit.dm:29]: Fixed!

File: code/modules/mob/living/carbon/human/species.dm [hunk 5]
```diff
@@ -612,14 +617,28 @@ GLOBAL_LIST_EMPTY(roundstart_races)
...
 					else
 						hair_overlay.color = "#" + H.hair_color
+
+					//Gradients
+					grad_style = H.grad_style
+					grad_color = H.grad_color
+					if(grad_style && grad_style != "None")
+						var/datum/sprite_accessory/gradient = GLOB.hair_gradients_list[grad_style]
+						var/icon/temp = new/icon("icon" = gradient.icon, "icon_state" = gradient.icon_state)
+						var/icon/temp_hair = new/icon("icon" = hair_file, "icon_state" = hair_state)
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm:627]: There's a cleaner way to make new icons
```suggestion
						var/icon/temp = icon(gradient.icon, gradient.icon_state)
						var/icon/temp_hair = icon(hair_file, hair_state)
```

File: code/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -574,6 +578,7 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 
 	if(!hair_hidden || dynamic_hair_suffix)
 		var/mutable_appearance/hair_overlay = mutable_appearance(layer = -HAIR_LAYER)
+		var/mutable_appearance/gradient_overlay = mutable_appearance(layer = -HAIR_LAYER)
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm:581]: Rather than a separate gradient overlay like this you could just be adding an overlay to the hair overlay

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by PKPenguin321** [comment]
More content for the pending Barber job I see

File: code/modules/mob/living/carbon/human/species.dm (hunk 6 above)
Comment (KathyRyals) [reply code/modules/mob/living/carbon/human/species.dm:581]: Discussed in coderbus and some oddities with MA makes this fine.

File: code/modules/mob/dead/new_player/sprite_accessories.dm [hunk 7]
```diff
@@ -779,6 +779,56 @@
...
+/////////////////////////////////////
+/  =---------------------------=    /
+/  == Gradient Hair Definitions ==  /
+/  =---------------------------=    /
+/////////////////////////////////////
+*/
+
+/datum/sprite_accessory/hair_gradient
+	icon = 'icons/mob/hair_gradients.dmi'
+/datum/sprite_accessory/hair_gradient/none
```
Comment (Fikou) [inline code/modules/mob/dead/new_player/sprite_accessories.dm:792]: make a newline between those

File: code/modules/mob/living/carbon/human/species.dm [hunk 8]
```diff
@@ -30,6 +30,10 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	///The alpha used by the hair. 255 is completely solid, 0 is invisible.
 	var/hair_alpha = 255
 
+	var/grad_style
+
+	var/grad_color
```
Comment (Fikou) [inline code/modules/mob/living/carbon/human/species.dm:35]: dm doc these

File: code/modules/mob/living/carbon/human/human_defines.dm (hunk 3 above)
Comment (Fikou) [inline code/modules/mob/living/carbon/human/human_defines.dm:23]: dm doc these

File: code/game/objects/items/dyekit.dm [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	icon_state = "dyespray"
+
+/obj/item/dyespray/attack_self(mob/user)
+	dye(user)
+
+/obj/item/dyespray/pre_attack(atom/target, mob/living/user, params)
+	dye(target)
+	return ..()
+
+/obj/item/dyespray/proc/dye(mob/target)
```
Comment (Fikou) [inline code/game/objects/items/dyekit.dm:14]: doc this

**Comment by KathyRyals** [comment]
Did the things.

File: code/modules/cargo/goodies.dm [hunk 10]
```diff
@@ -132,7 +132,11 @@
 	desc = "The most important thing you could possibly spend your hard-earned money on."
 	cost = PAYCHECK_EASY * 4
 	contains = list(/obj/item/toy/plush/beeplushie)
-
+/datum/supply_pack/goody/dyespray
```
Comment (ninjanomnom) [inline code/modules/cargo/goodies.dm:135]: Please put an empty line before the type like the rest of the file, same for the one below this.

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 11]
```diff
@@ -18,6 +18,11 @@
 	var/hair_color = "000"
 	var/hairstyle = "Bald"
 
+	///Colour used for the hair gradient.
+	var/grad_color = "000"
+	///Style used for the hair gradient.
+	var/grad_style = "None"
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/human_defines.dm:24]: Just have null be none

File: code/modules/mob/living/carbon/human/species.dm [hunk 12]
```diff
@@ -612,14 +617,28 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 							hair_overlay.color = "#" + hair_color
 					else
 						hair_overlay.color = "#" + H.hair_color
+
+					//Gradients
+					grad_style = H.grad_style
+					grad_color = H.grad_color
+					if(grad_style && grad_style != "None")
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm:624]: Once null is the none value you dont need the second check here

File: code/modules/vending/games.dm [hunk 13]
```diff
@@ -20,7 +20,8 @@
 					/obj/item/skillchip/bonsai=2,
 					/obj/item/skillchip/wine_taster=2,
 					/obj/item/skillchip/light_remover=2,
-					/obj/item/skillchip/useless_adapter=5)
+					/obj/item/skillchip/useless_adapter=5,
+					/obj/item/dyespray=3)
```
Comment (ninjanomnom) [inline code/modules/vending/games.dm:24]: Just a bit of cleanup while you're here, change this to a slightly better format like this:
```
listvar = list(
  foo,
  bar,
)
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 14]
```diff
@@ -18,6 +18,11 @@
 	var/hair_color = "000"
 	var/hairstyle = "Bald"
 
+	///Colour used for the hair gradient.
+	var/grad_color = "000"
+	///Style used for the hair gradient.
+	var/grad_style = null
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/human_defines.dm:24]: You don't need to specify null
```suggestion
	var/grad_style
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — APPROVED** [review]
